src/mirror.c

```diff
@@ -0,0 +1,233 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2018 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+#include <assert.h>
+#include <errno.h>
+#include <getopt.h>
+#include <libgen.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "swupd.h"
+
+static void print_help(const char *name)
+{
+	fprintf(stderr, "Usage:\n");
+	fprintf(stderr, "   swupd %s [options] [mirror_url]\n\n", basename((char *)name));
+	fprintf(stderr, "Help Options:\n");
+	fprintf(stderr, "   -h, --help              Show help options\n");
+	fprintf(stderr, "   -s, --set               set mirror url\n");
+	fprintf(stderr, "   -u, --unset             unset mirror url\n");
+	fprintf(stderr, "   -p, --path=[PATH...]    Use [PATH...] as the path to set/unset the mirror url for (eg: a chroot or btrfs subvol\n");
+	fprintf(stderr, "\n");
+}
+
+static const struct option prog_opts[] = {
+	{ "help", no_argument, 0, 'h' },
+	{ "set", required_argument, 0, 's' },
+	{ "unset", no_argument, 0, 'u' },
+	{ "path", required_argument, 0, 'p' },
+	{ 0, 0, 0, 0 }
+};
+
+static int unset_mirror_url()
+{
+	char *fullpath;
+	if (path_prefix != NULL) {
+		fullpath = mk_full_filename(path_prefix, "/etc/swupd");
+	} else {
+		string_or_die(&fullpath, "/etc/swupd");
+	}
+
+	assert(fullpath != NULL);
+	assert(strcmp(fullpath, "/") != 0);
+	assert(strcmp(fullpath, "/etc") != 0);
+	return swupd_rm(fullpath);
+}
+
+static int write_to_path(char *content, char *path)
+{
+	char *dir, *tmp = NULL;
+	string_or_die(&tmp, "%s", path);
+	dir = dirname(tmp);
+	/* attempt to  make the directory
+	 * ignore EEXIST errors */
+	int ret = mkdir(dir, S_IRWXU);
+	char *cmd;
+	string_or_die(&cmd, "mkdir -p %s", dir);
+	ret = system(cmd);
+	if (ret) {
+		free_string(&cmd);
+		free_string(&tmp);
+		return ret;
+	}
+	free_string(&cmd);
+	free_string(&tmp);
+
+	/* now try to open the file to write */
+	FILE *fp = NULL;
+	fp = fopen(path, "w");
+	if (fp == NULL) {
+		return 1;
+	}
+
+	/* and write to the file */
+	ret = fputs(content, fp);
+	if (ret < 0 || ret == EOF) {
```

Comment:
If we
      ret = fputs(content, fp);
      fclose(fp);
      if (ret < 0 || ret == EOF) { ....

we only need one fclose